### PR TITLE
Speedup x200 calc uv islands and add bbox module and apply new tools to Sort Align operator

### PR DESCRIPTION
**Pull request by Oxicid**
These are milestone changes, after which the addon will have fewer crutches, less code, and more speed of execution and fewer limitations.

Function get_selected_islands - as much as possible, effectively calculates islands, without requiring various manipulations with built-in operators.

Many functions related to bbox also need to be gradually moved to the appropriate module, as well as get rid of some code duplication.

The bbox module is very convenient to work with, there is no need to calculate (in vain) different values, they can be calculated through properties. And accessing these properties is more convenient than using dict and probably faster.

### CODE REVIEW

**Comment by franMarz** [comment]
First of all, thank you.
This is work in progress, right? Because a don't see many of the new functions being used across the code.
I was testing your get_selected_islands some time ago, I found it in one of your repos. In my tests, it took twice the time to finish the operation in a 5 times subdivided Suzanne (around 500k poligons). Maybe it performs well with low poly meshes, but I don't find it useful as we need the code to better scale with polycount.
The other improvements you bring have a good look, but I need to test them further.

**Comment by Oxicid** [comment]
@franMarz This PR is final, so that there are not too many additions, I'm waiting for you to accept these changes (I probably merged everything into a pile again, and these commits should have been made a separate PR, but there are a lot of related things).

About the miscalculation function of islands, yes on high-polygonal objects the difference is not significant, although it is there. But on smaller and more significant islands, the former method has exponential growth, on 5 thousand low poly islands gain of 2344 times, on middle poly 15-20, on high poly (250k faces) speed gain of 1.5 times.

The advantages of this function is that it can work in synchronization mode, which allows you to unblock some operators for this mode.
Also the former method needs to access bpy.ops very often, which updates mesh data with each access, which means that modifiers (if they are present) need to be updated every time, and here the calculation becomes terribly slow (2-20 times), and this is taking into account that the objects that I tested do not have so many attributes. Plus because of these updates, handlers are called, which sometimes causes serious performance problems.

Another plus is that 2 polygons connected by only one vertex, mo method counts for 2 islands (this behavior is typical for packers).

Perhaps my method uses less RAM (I haven't checked).

In short, only pluses.
[uv_calc_test.zip](https://github.com/franMarz/TexTools-Blender/files/14135278/uv_calc_test.zip)

The main thing that you accept this PR, so that I can further improve the addon

**Comment by franMarz** [comment]
- In the current state of the code, your get_selected_islands function is not used by any operator, make it the default method for island calculation where possible or delete it from the pull request, a commit shouldn't add unused code.
- Unwrap is misfiring for me when applied to any entirely selected island of Suzanne, rotating them out of position.
- Align_sort vertical now sorts the islands in the opposite direction compared to the previous version, is this intentional?
- In your added code, the indentation style don't match the one used in the rest of module.

The benefits of your code may be huge, but as the responsible of this add-on, I won't merge anything in a hurry. You can continue improving it in your fork if you want and open more pull requests after I eventually merge this. This is not a bugfix which should be handled quickly anyway.

**Comment by Oxicid** [comment]
@franMarz 
1) The get_selected_island function is used in the Sort Align operator, and I plan to continue using it (just not all at once, otherwise it will be difficult to fix potential bugs).

2) The second point I don't quite understand, do you by any chance use version 1.5, where the contex of the function bpy.ops.uv.cursor_set is redefined by the old type, I would like to see the file and know the exact steps of reproduction of this error. I have not changed the logic of the island_rotate function, it works fine for me.

3) Now the sorting direction corresponds to the icon, although I personally would like to flip the horizontal sorting icon, and align to the bottom edge. It's more intuitive that way.

4) Is this a pressing problem? I can change it no problem, after dealing with the issues above.

**Comment by franMarz** [comment]
I do all my test with the current version of the add-on from master, in Blender 4.0, in a plain file, only Suzanne. For stress tests, I apply some amount of subdivision to Suzanne.

In my tests, the current get islands function and your function have very similar timings and memory usage. I would like to see your function with all arguments from the current getSelectionIslands() added and in use, but your function is fine, not a x200 speedup but a better approach for sure:

0sub Suzanne 500 faces:     0.015s (patch) - 0.015s (master)
1sub Suzanne 1968 faces:    0.016s (patch) - 0.016s (master)
2sub Suzanne 7872 faces:    0.062s (patch) - 0.047s (master)
3sub Suzanne 31488 faces:   0.235s (patch) - 0.203s (master)
4sub Suzanne 125952 faces:  0.953s (patch) - 0.907s (master)
5sub Suzanne 503808 faces:  3.703s (patch) - 3.344s (master)
Total memory allocated:     813MB (patch) - 804MB (master)

We can adopt your function when finished or use it only in sort for now, I don't really care.

I can change the horizontal sort icon if you make your function to align to the bottom. I'll commit it after this is merged.
Nothing more for now. I hope you find out what's happening with that rotation function, otherwise I might look into it when I have time.

File: utilities_uv.py [hunk 1]
```diff
@@ -276,24 +278,30 @@ def move_island(island, dx, dy):
...
-			xt = x - center_x
-			yt = y - center_y
-			xr = (xt * math.cos(angle)) - (yt * math.sin(angle))
-			yr = (xt * math.sin(angle)) + (yt * math.cos(angle))
+			loop[uv_layer].uv += delta
+
+
+def rotate_island(island, uv_layer=None, angle=0, pivot=None):
+    '''Rotate a list of faces by angle (in radians) around a center'''
+    rot_matrix = mathutils.Matrix.Rotation(-angle, 2)
```
Comment (franMarz) [inline utilities_uv.py:289]: Unwrap and Sort tools work for me with angle instead of -angle. I'm using your code from master on a "Monkey" mesh.

**Review by franMarz — CHANGES_REQUESTED** [review]

**Comment by Oxicid** [comment]
@franMarz I suspect that your test does not take into account the time required to select faces and then restore them.

Yes on the monkey there is no strong difference, but you try editing mode duplicate the meshes that the number of islands was a few dozen, there will be a strong lag.

And when I meant about modifiers, I was talking about unapplied modifiers, just hover modifier subdive without applying and look at the result, there will be strong brakes.

And I have given above zip file with tests that you can use and make sure that the acceleration can reach even higher than x200.

1) And again, this function works also in synchronization mode
2) It doesn't need to pass selected faces and then restore selected faces at the end (less control).
3) It works easily in multi-object mode, no need to call multi_object_loop which greatly affects performance by calling frequent object updates.
4) Non-manifold islands are also skipped.
5) About RAM, still the old function consumes about 30 mb on 500k polygons, and the new 3-4 (the difference is not big and such an advantage for 2024).
6) This function can be implemented as a generator, which can avoid multiple memory reallocations that slow down python. 
7)And if you sort two separate objects in the former operator, the sorting will be "camel case", whereas in this PR it is linear. But this is not surprising, with multi_object_loop it is harder to achieve this result

**Comment by franMarz** [comment]
I understand your function is way better than the current one, specially in some complex scenarios. As I said, I want it to be usable for more operators with the addition of the remaining arguments the current function has. Anyway, we'll adopt your function as it is now, I hope you can generalize its use in the future.
